src/org/labkey/targetedms/view/spectrum/LibrarySpectrumMatchGetter.java

```diff
@@ -122,106 +119,97 @@
-                precursor.getModifiedSequence(),
-                precursor.getCharge());
-
-        return makeLibrarySpectrumMatch(spectrum, run, peptide, precursor, library);
-    }
-
-    public static LibrarySpectrumMatch getRedundantSpectrumMatch(TargetedMSRun run, Peptide peptide, Precursor precursor,
-                                                                 PeptideSettings.SpectrumLibrary library, Path redundantLibPath,
-                                                                 Container container, int redundantRefSpecId)
-    {
-        BlibSpectrum spectrum = BlibSpectrumReader.getRedundantSpectrum(container, redundantLibPath, redundantRefSpecId);
-
-        return makeLibrarySpectrumMatch(spectrum, run, peptide, precursor, library);
-    }
-
-    private static LibrarySpectrumMatch makeLibrarySpectrumMatch(BlibSpectrum spectrum, TargetedMSRun run, Peptide peptide, Precursor precursor,
-                                                                 PeptideSettings.SpectrumLibrary library)
+                                                        PeptideSettings.SpectrumLibrary library, Path libFilePath,
+                                                        Container container, int redundantRefSpectrumId, String sourceFile) throws SpeclibReaderException
     {
-        List<Peptide.StructuralModification> structuralModifications = ModificationManager.getPeptideStructuralModifications(precursor.getGeneralMoleculeId());
-        List<PeptideSettings.RunStructuralModification> runStrMods = ModificationManager.getStructuralModificationsForRun(run.getId());
-        Map<Long, List<PeptideSettings.PotentialLoss>> potentialLossMap = new HashMap<>();
-        for(Peptide.StructuralModification mod: structuralModifications)
+        String fileName = FileUtil.getFileName(libFilePath);
+        LibSpectrumReader reader = getReaderForLibrary(fileName);
+        if(reader != null)
         {
-            List<PeptideSettings.PotentialLoss> losses = ModificationManager.getPotentialLossesForStructuralMod(mod.getStructuralModId());
-            potentialLossMap.put(mod.getStructuralModId(), losses);
-        }
-
-        LibrarySpectrumMatch pepSpec = null;
-
-        // Make sure that the Bibliospec spectrum has peaks.  Minimized libraries in Skyline can have
-        // library entries with no spectrum peaks.  This should be fixed in Skyline.
-        if(spectrum != null && spectrum.getNumPeaks() > 0)
-        {
-            pepSpec = makeLibrarySpectrumMatch(spectrum, peptide, precursor, library, structuralModifications, runStrMods, potentialLossMap);
+            LibSpectrum spectrum = null;
+            try
+            {
+                if (reader instanceof BlibSpectrumReader)
+                {
+                    spectrum = redundantRefSpectrumId != 0 ? ((BlibSpectrumReader) reader).getRedundantSpectrum(container, libFilePath, redundantRefSpectrumId)
```

Comment:
Any way this could be pushed into a method on the readers instead of having to do instanceof checks and cast?

Reply:
Done